kvdb/cache.go

```diff
@@ -0,0 +1,930 @@
+}
+
+// Enforce that Cache implements the ExtendedBackend interface.
+var _ walletdb.DB = (*Cache)(nil)
+
+type pendingChange interface {
+	// Reverts the pending change.
+	Revert()
+}
+
+type pendingAdd struct {
+	parent   *cachedItem
+	newChild *cachedItem
+}
+
+func (p *pendingAdd) Revert() {
+	p.parent.bucket.tree.Delete(p.newChild)
+}
+
+type pendingUpdate struct {
+	parent   *cachedItem
+	oldValue string
+}
+
+func (p *pendingUpdate) Revert() {
+	p.parent.value = p.oldValue
+}
+
+type pendingDelete struct {
+	parent   *cachedItem
+	oldChild *cachedItem
+}
+
+func (p *pendingDelete) Revert() {
+	p.parent.bucket.tree.ReplaceOrInsert(p.oldChild)
+}
+
+// Cache is a simple write through cache implementing the kvdb.Backend
+// interface. It's capable of recursively caching top-level buckets speeding up
+// reads by reducing roundtrips to the actual DB backend while remaining
+// consistend with DB state as long as the cache is used when mutating those
+// buckets.
+type Cache struct {
+	mx       sync.RWMutex
+	currRwTx *cacheReadWriteTx
+	root     *cachedItem
+	pending  *list.List
+	backend  Backend
+}
+
+// NewCache constructs a new cache.
+func NewCache(backend Backend) *Cache {
+	return &Cache{
+		root: &cachedItem{
+			bucket: newCacheBucket(),
+		},
+		pending: list.New(),
+		backend: backend,
+	}
+}
+
+func (c *Cache) pendingAdd(parent *cachedItem, newChild *cachedItem) {
+	c.pending.PushBack(&pendingAdd{parent, newChild})
+}
+
+func (c *Cache) pendingUpdate(parent *cachedItem, oldValue string) {
+	c.pending.PushBack(&pendingUpdate{parent, oldValue})
+}
+
+func (c *Cache) pendingDelete(parent *cachedItem, oldChild *cachedItem) {
+	c.pending.PushBack(&pendingDelete{parent, oldChild})
+}
+
+type traversalHelper struct {
+	bucket walletdb.ReadBucket
+	root   *cachedItem
+}
+
+// AddTopLevelBucket recursively reads the passed top-level bucket and adds
+// all content below it to the cache.
+func (c *Cache) AddTopLevelBucket(key []byte) error {
+	c.mx.Lock()
+	defer c.mx.Unlock()
+
+	var root *cachedItem
+
+	err := View(c.backend, func(tx RTx) error {
+		var queue []traversalHelper
+
+		currBucket := tx.ReadBucket(key)
+		if currBucket == nil {
+			return nil
+		}
+
+		root = &cachedItem{
+			key:    string(key),
+			bucket: newCacheBucket(),
+		}
+
+		currRoot := root
+
+		for {
+			currBucket.ForEach(func(k, v []byte) error {
+				var item *cachedItem
+				if v == nil {
+					item = &cachedItem{
+						key:    string(k),
+						bucket: newCacheBucket(),
+					}
+
+					bucket := currBucket.NestedReadBucket(k)
+					queue = append(
+						queue,
+						traversalHelper{
+							bucket: bucket,
+							root:   item,
+						},
+					)
+				} else {
+					item = &cachedItem{
+						key:   string(k),
+						value: string(v),
+					}
+
+				}
+
+				currRoot.bucket.put(item)
+				return nil
+			})
+
+			if len(queue) == 0 {
+				break
+			}
+
+			currBucket = queue[0].bucket
+			currRoot = queue[0].root
+			queue = queue[1:]
```

Comment:
If we do `queue[0] = nil` here, then it's possible to incrementally free up more memory as the algo executes.